### DISCUSSION TITLE
关于配置文件的配置方式

### DISCUSSION THREAD

**Discussion by mrzkj** [rathole-org/rathole#215] [Ideas]
rathole的配置文件能否更改为单侧配置？而不是每增加一个端口，需要去动两侧的配置文件？
比如说frp的配置文件，虽然服务端也需要配置，但是服务端的配置文件属于一次配置，终生使用，增加新的端口无需再进入服务端进行配置，只需要在客户端增加新的配置条件即可。
再比如说nps的配置文件，是在服务端使用的，同时也支持只客户端进行配置，增加新的端口的时候无需配置客户机和服务器两侧的配置文件。
rathole能否也进行升级，增加新的端口的时候只在服务端或者客户端的某一侧进行配置即可，而不是两侧都要进行配置？谢谢！

**Comment by mrzkj** [comment]
说实话我个人更偏向将配置文件放在客户端，因为这样如果需要更换穿透主机的时候，只需要在新的穿透主机上全新配置一个简单的服务端即可，不需要从旧的服务器上转移配置文件了。

**Comment by Envov** [comment]
可以为服务端提供一个面板，客户端还是只配置协议个端口的简短信息比较好

**Comment by konanzheng** [comment]
假设A, C两个设备都在各自内网中，都需要外网访问8000端口，通过服务器 B,进行端口映射，A, C都可以通过B的公网IP访问到自己的8000端口。下面这个描述整个端口配置过程。


A或C 客户端配置，配服务器 ip,端口 和token，以及本地端口(支持多个本地端口以及局域网配置)
```
[client]
remote_addr = "myserver.com:2333"
[client.services.my8000]
token = "use_a_secret_that_only_you_know"
local_addr = "127.0.0.1:8000"
```
服务器配置,不针对每个客户端配置，只配端口范围
```
[server]
bind_addr = "0.0.0.0:2333"

[server.services.anything]
token = "use_a_secret_that_only_you_know"
ports = "3000-4000,5000-6000,8000-28000" #服务器可用端口范围，自动分配可用端口，优先使用客户端相同端口
```


最终结果
1. 服务器IP:8000 映射到 A的8000端口
2. 服务器IP:28000 映射到 C的8000端口

```mermaid
sequenceDiagram
    participant A as device A
    participant B as public Server
    participant C as device C
    A ->> B: request for share my port 8000
    Note right of B: check token and port
    B -->> A: ok, server-ip:8000 is ready
    Note over A,B : device A 8000 is public by server-ip:8000
    C ->> B: request for share my port 8000
    Note left of B: check token and port
    B -->> C:8000 is busy,use server-ip:28000
    Note over B,C: connect to C 8000 by server-ip:28000
```
